conventions/cli.md

```diff
@@ -1,79 +1,181 @@
+- `create` creates a resource.
+- `update` updates a resource.
+- `delete` deletes a resource.
 
-- `delete` deletes a resource
+For a given resource, create and update should use the same arguments as much as possible and where it makes sense.
 
-For a given resource there should be parallelism between arguments to `create`
-and `update` as much as possible.
+Other domain-specific verbs are possible on a case-by-case basis for operations that go beyond basic CRUD operations.
 
-Other domain-specific verbs are possible on a case-by-case basis, like
-`set-traffic` for a Knative Serving Service.
+### Identifier
 
-## Arguments
+For the `CRUD` operations `describe`, `create`, `update`, `delete` the identifier is the resource's name and is required as a positional argument after the commands.
+For example it is the last argument that does not start with a flag prefix `-` or `--`.
+`list` operations can use a resource name to filter on the resource.
 
-### Positionals
+Other identifiers can be plugin names or other entities' identifiers.
 
-Where there's a single target resource, the resource name should be a positional
-argument. It needs to be of the resource type we're talking about, eg.
-`kn revision` subcommands the positional must be naming a revision.
+For bulk operations also multiple identifiers can be provided.
+For example, a `delete` operation could use multiple resource names that should be deleted.
 
 ```bash
 kn service create foo --image gcr.io/things/stuff:tag
 ```
 
-In this case `foo` is positional, and refers to the service to create.
+In this case, `foo` is positional, and provides the name of the service to create.
 
 ### Flags
 
-- `--force` is a flag on all create commands, and will replace the resource if
-  it already exists (otherwise this is an error). The resource will be _mutated_
-  to have a spec exactly like the resource that would otherwise be created. It
-  is not deleted and recreated.
+Flags are used for specifying the input for `kn` commands and can have different characteristics:
+
+* They can be _mandatory_ or _optional_
+* Optional flags can have _default values_
+* Flag values can be _scalars_, _binary_, _lists_ or _maps_ (see below for details)
+* Flags always have a long-form (starting with a double `--`) but can also have a shortcut (beginning with a single `-`)
+* Every flag has a help message attached
+* Flags can be specific to a command or can be globally applicable
+
+When adding new flags, the following recommendations should be considered:
+
+* Never add a global flag except for very good reasons
+* Group related flags together by using a common prefix, like `--label-revision` or `--label-service` so that they appear together in the help message (which is sorted alphabetically)
+* Don't add a short form without former discussions
+* Choose a name for the flag that is the same or close to the naming used in Knative serving itself like the corresponding CRD field or annotation name.
```

Comment:
<!--mdfmt-bot-->
Format markdown:
```suggestion
- Never add a global flag except for very good reasons
- Group related flags together by using a common prefix, like `--label-revision`
  or `--label-service` so that they appear together in the help message (which
  is sorted alphabetically)
- Don't add a short form without former discussions
- Choose a name for the flag that is the same or close to the naming used in
  Knative serving itself like the corresponding CRD field or annotation name.
```